### PR TITLE
Support nv15

### PR DESCRIPTION
**Pull request by Stebalien**
Rebased for builtin-actors extraction.

Closes https://github.com/filecoin-project/ref-fvm/issues/335.

### CODE REVIEW

**Comment by arajasek** [comment]
Closes #335

File: fvm/src/machine/default.rs [hunk 1]
```diff
@@ -56,26 +58,29 @@ where
...
         base_fee: TokenAmount,
-        fil_vested: TokenAmount,
+        circ_supply: TokenAmount,
         network_version: NetworkVersion,
         state_root: Cid,
         builtin_actors: Cid,
         blockstore: B,
         externs: E,
     ) -> anyhow::Result<Self> {
+        const SUPPORTED_VERSIONS: RangeInclusive<NetworkVersion> =
```
[the remark this answers is not in the document]
Comment (arajasek) [reply fvm/src/machine/default.rs:68]: From offline sync: we will try to support both v14 and v15, that will require adjustments to this changeset

File: fvm/src/gas/price_list.rs [hunk 2]
```diff
@@ -382,6 +384,11 @@ impl PriceList {
             0,
         )
     }
+    /// Returns gas required for replica verification.
+    #[inline]
+    pub fn on_verify_replica_info(&self, _replica: &ReplicaUpdateInfo) -> GasCharge<'static> {
```
Comment (Stebalien) [inline fvm/src/gas/price_list.rs:389]: `on_verify_replica_update`?

File: fvm/src/kernel/default.rs [hunk 3]
```diff
@@ -399,13 +401,17 @@ where
...
     fn total_fil_circ_supply(&self) -> Result<TokenAmount> {
-        Ok((&self.call_manager.context().fil_vested
-            + &self.get_fil_mined()?
-            + &self.get_reserve_disbursed()?
-            - &self.get_burnt_funds()?
-            - &self.power_locked()?
-            - &self.market_locked()?)
-            .max(Zero::zero()))
+        if self.network_version() <= V14 {
+            return Ok((&self.call_manager.context().circ_supply
```
Comment (Stebalien) [inline fvm/src/kernel/default.rs:405]: comment explaining wtf

File: fvm/src/machine/default.rs [hunk 4]
```diff
@@ -196,41 +203,84 @@ where
     }
 
     fn transfer(&mut self, from: ActorID, to: ActorID, value: &TokenAmount) -> Result<()> {
-        if from == to || value.is_zero() {
-            return Ok(());
-        }
-
-        if value.is_negative() {
-            return Err(syscall_error!(IllegalArgument;
+        if self.context.network_version >= V15 {
```
Comment (Stebalien) [inline fvm/src/machine/default.rs:206]: `NetworkVersion::V15` is easier to understand.

File: fvm/src/machine/default.rs [hunk 5]
```diff
@@ -196,41 +203,84 @@ where
     }
 
     fn transfer(&mut self, from: ActorID, to: ActorID, value: &TokenAmount) -> Result<()> {
```
Comment (Stebalien) [inline fvm/src/machine/default.rs:205]: can we somehow reduce the code duplication here?

File: shared/src/sector/registered_proof.rs [hunk 6]
```diff
@@ -87,6 +87,12 @@ impl RegisteredSealProof {
     }
 }
 
+impl Default for RegisteredSealProof {
+    fn default() -> Self {
+        Self::Invalid(0)
```
Comment (Stebalien) [inline shared/src/sector/registered_proof.rs:92]:
```suggestion
        Self::Invalid(-1)
```

File: testing/conformance/Cargo.toml [hunk 7]
```diff
@@ -44,7 +44,7 @@ serde_json = { version = "1.0", features = ["raw_value"] }
 walkdir = "2.3"
 regex = { version = "1.0" }
 ittapi-rs = { version = "0.1.6", optional = true }
-actors-v6 = { package = "fil_builtin_actors_bundle", git = "https://github.com/filecoin-project/builtin-actors", branch = "v6" }
+actors-v6 = { package = "fil_builtin_actors_bundle", git = "https://github.com/filecoin-project/builtin-actors", branch = "release/v6" }
```
Comment (Stebalien) [inline testing/conformance/Cargo.toml:47]: Did you create this branch?

**Review by Stebalien — COMMENTED** [review]
Some nits, otherwise lgtm.

File: fvm/src/gas/price_list.rs (hunk 2 above)
Comment (arajasek) [reply fvm/src/gas/price_list.rs:389]: Yessir

File: fvm/src/kernel/default.rs (hunk 3 above)
Comment (arajasek) [reply fvm/src/kernel/default.rs:405]: yessir

File: fvm/src/machine/default.rs (hunk 4 above)
Comment (arajasek) [reply fvm/src/machine/default.rs:206]: yessir

File: fvm/src/machine/default.rs (hunk 5 above)
Comment (Stebalien) [reply fvm/src/machine/default.rs:205]: resolved: eh, it's simpler this way.

File: fvm/src/kernel/default.rs [hunk 8]
```diff
@@ -399,13 +401,17 @@ where
...
 {
     fn total_fil_circ_supply(&self) -> Result<TokenAmount> {
-        Ok((&self.call_manager.context().fil_vested
-            + &self.get_fil_mined()?
-            + &self.get_reserve_disbursed()?
-            - &self.get_burnt_funds()?
-            - &self.power_locked()?
-            - &self.market_locked()?)
-            .max(Zero::zero()))
+        if self.network_version() <= V14 {
```
Comment (raulk) [inline fvm/src/kernel/default.rs]:
```suggestion
        if self.network_version() <= NetworkVersion::V14 {
```

File: testing/conformance/Cargo.toml (hunk 7 above)
Comment (arajasek) [reply testing/conformance/Cargo.toml:47]: Yes i did

File: fvm/src/kernel/default.rs (hunk 8 above)
Comment (raulk) [reply fvm/src/kernel/default.rs]: Would probably prefer something like this for a single return site:

```rust
let circ_supply = if self.network_version() <= NetworkVersion::V14 {
    ...
} else {
    self.call_manager.context().circ_supply.clone()
};
Ok(circ_supply)
```

File: shared/src/version/mod.rs [hunk 9]
```diff
@@ -39,6 +39,8 @@ pub enum NetworkVersion {
     V13,
     /// actors v6
     V14,
+    /// actors v6
```
Comment (raulk) [inline shared/src/version/mod.rs:42]:
```suggestion
    /// actors v7
```

File: testing/conformance/Cargo.toml (hunk 7 above)
Comment (raulk) [reply testing/conformance/Cargo.toml:47]: This is the branch name we converged to for long-term maintenance of release lines.

**Review by raulk — APPROVED** [review]

File: shared/src/version/mod.rs (hunk 9 above)
Comment (arajasek) [reply shared/src/version/mod.rs:42]: eagle eyes

**Pull request merged**